### ISSUE TITLE
CSI attacher couldn't redeploy when Kubernetes node is down

### ISSUE THREAD

**Issue by JacieChao** [rancher/longhorn#199]
Environment: Using rke to deploy kubernetes v1.10.5

Test steps:
1. Deploy longhorn with CSI driver
2. Power off CSI attacher pod nodes.
3. The driver deployer couldn't restart, always with `CrashLoopBackOff` status. and csi-attacher pod couldn't schedule to other nodes.
Here's the driver deployer logs:
```
time="2018-08-09T15:34:39Z" level=debug msg="Driver csi will be used after automatic detection"
time="2018-08-09T15:34:39Z" level=debug msg="Deploying CSI driver"
time="2018-08-09T15:34:39Z" level=debug msg="Trying to get the service csi-attacher"
time="2018-08-09T15:34:39Z" level=debug msg="Got the service csi-attacher"
time="2018-08-09T15:34:39Z" level=debug msg="Trying to delete the service csi-attacher"
time="2018-08-09T15:34:39Z" level=debug msg="Deleted the service csi-attacher"
time="2018-08-09T15:34:39Z" level=debug msg="Waiting for foreground deletion of service csi-attacher"
time="2018-08-09T15:34:40Z" level=debug msg="Deleted the service csi-attacher in foreground"
time="2018-08-09T15:34:40Z" level=debug msg="Trying to create the service csi-attacher"
time="2018-08-09T15:34:40Z" level=debug msg="Created the service csi-attacher"
time="2018-08-09T15:34:40Z" level=debug msg="Trying to get the statefulset csi-attacher"
time="2018-08-09T15:34:40Z" level=debug msg="Got the statefulset csi-attacher"
time="2018-08-09T15:34:40Z" level=debug msg="Trying to delete the statefulset csi-attacher"
time="2018-08-09T15:34:40Z" level=debug msg="Deleted the statefulset csi-attacher"
time="2018-08-09T15:34:40Z" level=debug msg="Waiting for foreground deletion of statefulset csi-attacher"
```

**Comment by niusmallnan** [comment]
This happens when the csi-attacher pod and the driver-deployer pod are on the same host.

If only csi-attacher pod has a problem, k8s can complete self-healing.
But the driver-deployer starts, it will delete the csi-attacher statefulset. This makes the situation a bit more complicated. Statefulset is really slow.

**Comment by yasker** [comment]
@niusmallnan In fact for the statefulset, I think k8s may not able self-healing by itself.

https://github.com/kubernetes/kubernetes/issues/47230#issuecomment-309586908

I've tried it once. A normal stateful set pod stuck in the `unknown` and nothing will happen until I delete the pod.

Though in the CSI case, even delete the pod doesn't work for me. Feels we're at some weird K8S corner case here.

**Comment by TehTotalPwnage** [comment]
I did some investigating regarding other implementations of `CSI` to see how they handle rescheduling of the `csi-attacher` and `csi-provisioner` (`csi-provisioner` is also faced with this issue because it is also deployed as a `StatefulSet`):
- Rook has [yet to develop](https://github.com/rook/rook/issues/1385) their [CSI implementation](https://github.com/rook/rook/projects/10).
- OpenEBS is [in the midst](https://github.com/openebs/openebs/issues/357) of [developing their implementation](https://github.com/openebs/csi-openebs), but their code has [no special implementation](https://github.com/openebs/csi-openebs/pull/4) for `csi-attacher` and `csi-provisioner`. They're using their deployment spec to create the `StatefulSets` instead. 
- [Ceph](https://github.com/ceph/ceph-csi/blob/master/deploy/cephfs/kubernetes/csi-cephfsplugin-attacher.yaml), [NFS](https://github.com/kubernetes-csi/drivers/blob/master/pkg/nfs/deploy/kubernetes/csi-attacher-nfsplugin.yaml), [DigitalOcean](https://github.com/digitalocean/csi-digitalocean/blob/master/deploy/kubernetes/releases/csi-digitalocean-v0.1.3.yaml), and [Google Compute Engine](https://github.com/kubernetes-sigs/gcp-compute-persistent-disk-csi-driver/blob/master/deploy/kubernetes/controller.yaml) are also simply using `YAML` files.
- [Portworx](http://install.portworx.com/1.4/) and [StorageOS](https://github.com/storageos/helm-chart/blob/master/templates/statefulset_csi.yaml) are both closed source, but have `YAML` files holding these containers, suggesting they also have no special handling of this case.

If we currently want to handle it using `StatefulSet`, we'd have to work around the stuck `Pod` by using [force deletion](https://kubernetes.io/docs/tasks/run-application/force-delete-stateful-set-pod/#delete-pods), but that's dangerous for [a number of reasons](https://kubernetes.io/docs/tasks/run-application/force-delete-stateful-set-pod/#statefulset-considerations).

Thankfully, it looks like the `Kubernetes` team has noted this issue as well. They've created a [Pod election specification](https://github.com/kubernetes/contrib/tree/master/election) that would allow for running of multiple `Pods` as fail-safes while also allowing for one specific `Pod` to be the "leader". They've [implemented an option to use leader elections](https://github.com/kubernetes-csi/external-attacher/blob/master/deploy/kubernetes/deployment.yaml#L114) in the [latest release](https://github.com/kubernetes-csi/external-attacher/releases/tag/v0.3.0) for `csi-attacher`, allowing us to run it as a `Deployment` instead and keep the `csi-attacher` online. However, the `csi-provisioner` hasn't implemented this yet, so we'll have to wait and see where this goes before implementing a solution to this issue.

**Comment by yasker** [comment]
Thanks @TehTotalPwnage . That's a good research.

@niusmallnan Can you check that if a normal statefulset works for you? I think I should reach out to SIG storage about this if we've confirmed that the culprit is the statefulset.

**Comment by niusmallnan** [comment]
@yasker Statefulset really can't complete self-healing, even if I have added a livenessprobe.

If I delete the dead pod with this: `kubectl delete pod/xxx --grace-period=0 --force`, this pod can be rescheduled and run

**Comment by niusmallnan** [comment]
If we shutdown one node which runs driver-deployer, driver-deployer will try to rebuild on other nodes.  The driver-deployer will rebuild soon because it is a deployment.

Whether it is CSI or Flexvolume, we now delete and create the corresponding spec.
For CSI, we need to completely remove two Statefulsets and one Daemonset.
For Flexvolume, only one Daemonset needs to be removed.
I believe that if the number of nodes is large, the time required to rebuild Daemonset will be longer.
During this time, the entire k8s cluster will not be able to obtain longhorn volumes.
In other words, the user only shutdown one host, but it does affect the entire cluster.
I don't think this is a good experience.

So I think the driver-deployer's logic should be like this:
1. Create driver spec only on first boot, and get rid of the `Delete` action. There will be no more `Delete` actions.
2. For CSI, every time the driver-deployer starts, we need to try to delete the exception pods of the Statefulset(e.g. kubectl delete pod/xxx --grace-period=0 --force).  Let k8s itself go to restore the Statefulset.

**Comment by yasker** [comment]
We would still need some way to clean up after the deployer. So I think the `Delete` is still needed.

But if we've detected the existing stateful set, we can check if it's the same setup as we expected. If so, we can skip the deletion and remove the exception pods (if any) only. Otherwise, we would need to remove the existing stateful set and deploy the new ones.

**Comment by yasker** [comment]
Push to v0.3.1 due to time constraint for the v0.3.0 release.

**Comment by yasker** [comment]
Filed https://github.com/kubernetes-csi/external-provisioner/issues/130

**Comment by yasker** [comment]
Fixed in https://github.com/rancher/longhorn-manager/commit/ba3478302db6c7cb1c207933294a42a73dc20809
